### PR TITLE
add jit post

### PR DESCRIPTION
**Pull request by skeydan**

### CODE REVIEW

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+
+#### The `torch` just-in-time compiler
+
+Compared to that notion of JIT, at once generic (in technical regard) and specific (in time), what (Py-)Torch people have in mind when they talk of "the JIT" is both more narrowly-defined (in terms of operations) and more inclusive (in time): What is understood is the complete process from providing code input that can be converted into an intermediate representation (IR), via generation of that IR, via successive optimization of the same by the JIT compiler, via conversion (again, by the compiler) to bytecode, to -- finally -- execution, again taken care of by that same compiler, that now is acting as a virtual machine.
+
+If that sounded complicated, don't be scared. To actually make use of this feature from R, not much is to be learned in terms of syntax; a single function, augmented by a few specialized helpers, is stemming all the heavy load. What matters though is to understand a bit about how JIT compilation works, so you know what to expect, and are not surprised by unintended outcomes.
```
Comment (batpigandme) [inline _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:43]:
```suggestion
If that sounded complicated, don't be scared. To actually make use of this feature from R, not much needs to be learned in terms of syntax; a single function, augmented by a few specialized helpers, is stemming all the heavy load. What matters, though, is understanding a bit about how JIT compilation works, so you know what to expect, and are not surprised by unintended outcomes.
```

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+
+## What's coming (in this text)
+
+This post has three further parts.
+
+In the first, we explain how to make use of JIT capabilities in R `torch`. Beyond the syntax, we focus on the semantics: what essentially happens when you "JIT trace" a piece of code, and how that affects the outcome.
```
Comment (batpigandme) [inline _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:49]:
```suggestion
In the first, we explain how to make use of JIT capabilities in R `torch`. Beyond the syntax, we focus on the semantics (what essentially happens when you "JIT trace" a piece of code), and how that affects the outcome.
```

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+
+In the second, we' re allowing ourselves to "peek under the hood" a little bit; feel free to just cursorily skim if this does not interest you too much.
```
Comment (batpigandme) [inline _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:51]:
```suggestion
In the second, we "peek under the hood" a little bit; feel free to just cursorily skim if this does not interest you too much.
```

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+
+In the third, we show an example of using JIT compilation to enable deployment in an environment that has no R installed.
```
Comment (batpigandme) [inline _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:53]:
```suggestion
In the third, we show an example of using JIT compilation to enable deployment in an environment that does not have R installed.
```

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+
+## How to make use of `torch` JIT compilation
+
+In Python-world, or more specifically, in Python incarnations of deep learning frameworks, there is a magic verb "trace" that refers to a way of obtaining a graph representation from executing code eagerly. Namely, you run a piece of code -- a function, say, containing PyTorch operations -- on example inputs. These example inputs are arbitrary value-wise, but (naturally) need to conform to the shapes expected by the function. Tracing will then record operations as executed, meaning: those operations that *were* in fact executed, and only those. Any code paths never entered are consigned to oblivion.
```
Comment (batpigandme) [inline _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:57]:
```suggestion
In Python-world, or more specifically, in Python incarnations of deep learning frameworks, there is a magic verb "trace" that refers to a way of obtaining a graph representation from executing code eagerly. Namely, you run a piece of code -- a function, say, containing PyTorch operations -- on example inputs. These example inputs are arbitrary value-wise, but (naturally) need to conform to the shapes expected by the function. Tracing will then record operations as executed, meaning: those operations that *were* in fact executed, and only those. Any code paths not entered are consigned to oblivion.
```

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+  
+  
+  x + d 
+  
+}
+
+
+```
+
+Here computation of `e` is useless -- it is never used. Consequently, in the intermediate representation, `e` does not even appear. Also, as the values of `a`, `b`, and `c` are known already at compile time, the only constant present in the IR is `d`, their sum.
```
Comment (batpigandme) [inline _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:141]: Nice explanation

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+
+Under certain conditions, several operations can be chained (or *fused*, to use the technical term) into a single one. Here, three of those four methods (namely, all but `torch_matmul()`) operate *point-wise*; that is, they modify each element of a tensor in isolation. In consequence, not only do they lend themselves optimally to parallelization individually, -- the same would be true of a function that were to *compose* ("fuse") them: To compute a composite function "multiply then add then ReLU"
+
+$$
+relu() \ \circ \ (+) \ \circ \ (*)
+$$
+
+on a tensor *element*, nothing needs to be known about other elements in the tensor. The aggregate operation could then be run on the GPU in a single kernel.
+
+To make this happen, you normally would have to write custom CUDA code. Thanks to the JIT compiler, in many cases you don't have to: It will create such a kernel on the fly*.*
```
Comment (batpigandme) [inline _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:189]: Are you emphasizing the period on this sentence on purpose? If so, I'm not sure I understand why.

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+    with prim::TensorExprGroup_0 = graph(%x.1 : Float(5, 5, strides=[5, 1], requires_grad=0, device=cuda:0)):
+      %4 : int = prim::Constant[value=1]()
+      %3 : Float(5, 5, strides=[5, 1], requires_grad=0, device=cuda:0) = prim::Constant[value=<Tensor>]()
+      %7 : Float(5, 5, strides=[5, 1], requires_grad=0, device=cuda:0) = prim::Constant[value=<Tensor>]()
+      %x.10 : Float(5, 5, strides=[5, 1], requires_grad=0, device=cuda:0) = aten::mul(%x.1, %7) # <stdin>:4:0
+      %x.6 : Float(5, 5, strides=[5, 1], requires_grad=0, device=cuda:0) = aten::add(%x.10, %3, %4) # <stdin>:5:0
+      %x.2 : Float(5, 5, strides=[5, 1], requires_grad=0, device=cuda:0) = aten::relu(%x.6) # <stdin>:6:0
+      return (%x.2)
+
+From this output, we learn that three of the four operations have been grouped together, to form a `TensorExprGroup` . This `TensorExprGroup` will be compiled into a single CUDA kernel. The matrix multiplication, however -- not being a pointwise operation -- has to be executed by itself.
```
Comment (batpigandme) [inline _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:222]:
```suggestion
From this output, we learn that three of the four operations have been grouped together to form a `TensorExprGroup` . This `TensorExprGroup` will be compiled into a single CUDA kernel. The matrix multiplication, however -- not being a pointwise operation -- has to be executed by itself.
```

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+
+At this point, we stop our exploration of JIT optimizations, and move on to the last topic: model deployment in R-less environments. If you'd like to know more, Thomas Viehmann's [blog](https://lernapparat.de/){.uri} has posts that go into incredible detail on (Py-)Torch JIT compilation.
+
+## `torch` without R
+
+Our plan is the following: We define and train a model, in R. Then, we trace and save it. The saved file is then `jit_load()`ed in another environment, an environment that does not have R installed. Any language that has an implementation of Torch will do, provided that implementation includes the JIT functionality. The most straightforward way to show how this works is using Python; for deployment with C++, please see the [detailed instructions](https://pytorch.org/tutorials/advanced/cpp_export.html) on the PyTorch website.
```
Comment (batpigandme) [inline _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:228]:
```suggestion
Our plan is the following: We define and train a model, in R. Then, we trace and save it. The saved file is then `jit_load()`ed in another environment, an environment that does not have R installed. Any language that has an implementation of Torch will do, provided that implementation includes the JIT functionality. The most straightforward way to show how this works is using Python. For deployment with C++, please see the [detailed instructions](https://pytorch.org/tutorials/advanced/cpp_export.html) on the PyTorch website.
```

**Review by batpigandme — APPROVED** [review]
Nice work!

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+    Epoch: 9, loss: 1.2824
+
+    Epoch: 10, loss: 1.2596
+
+#### Trace in `eval` mode
+
+Now, for deployment, we want a model that has the same weights as the model we just trained, but does *not* drop out any tensor elements. A way to achieve is to instantiate a fresh model instance, immediately put it into `eval()` mode, trace it, and *then only* endow it with the trained model's weights.
+
+```{r}
+deploy_model <- net()$eval()
```
Comment (dfalbel) [inline _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:360]: Since you are not training a traced model, I think you could simply do `train_model$eval()` and then `jit_trace(train_model, ...)`. No need to load the weights again or trace a clean `net()`. You would only need that if `train_model()` was also traced becaus in that case `train_model$eval()` would have no effect in disabling the dropout.

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+      torch_randn(n, 1)
+    
+  },
+  
+  .getitem = function(i) {
+    list(x = self$x[i, ], y = self$y[i])
+  },
+  
+  .length = function() {
+    self$x$size()[[1]]
```
Comment (dfalbel) [inline _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:288]:
```suggestion
    self$x$size(1)
```

**Review by dfalbel — APPROVED** [review]
Looks great @skeydan ! Really liked the intro!

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+preview: images/preview.jpg
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+*Note: To follow along with this post, you will need `torch` version 0.5, which as of this writing is not yet on CRAN. In the meantime, please install the development version from GitHub.*
+
+Every domain has its concepts, and these are what one needs to understand, at some point, on one's journey from copy-and-make-it-work to purposeful, deliberate utilization. In addition, unfortunately, every domain has its jargon, whereby terms are used in a way that is technically correct, but fails to evoke a clear image to the yet-uninitiated. (Py-)Torch's JIT is an example.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:27]: I meant the latter ... as opposed to what one might think  of if coming from, e.g., java

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd (hunk 2 above)
Comment (skeydan) [reply _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:49]: haha the usual (colon ;-))
guess i need to not be stubborn at least once in a while :-))

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd (hunk 5 above)
Comment (skeydan) [reply _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:57]: oh, great catch. "never" was dramatic, but actually, illogical!

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd (hunk 7 above)
Comment (skeydan) [reply _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:189]: oops, idk how that happened ...? will remove

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd (hunk 3 above)
Comment (skeydan) [reply _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:51]: got it!

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+    self_contained: false
+    toc: true
+preview: images/preview.jpg
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+*Note: To follow along with this post, you will need `torch` version 0.5, which as of this writing is not yet on CRAN. In the meantime, please install the development version from GitHub.*
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:25]: done!

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+
+Every domain has its concepts, and these are what one needs to understand, at some point, on one's journey from copy-and-make-it-work to purposeful, deliberate utilization. In addition, unfortunately, every domain has its jargon, whereby terms are used in a way that is technically correct, but fails to evoke a clear image to the yet-uninitiated. (Py-)Torch's JIT is an example.
+
+## Terminological introduction
+
+"The JIT", much talked about in PyTorch-world and an eminent feature of R `torch`, as well, is two things at the same time -- depending on how you look at it: an optimizing compiler; and a free pass to execution in many environments where neither R nor Python are present.
+
+#### Compiled, interpreted, just-in-time compiled
+
+Starting with the former: "JIT" is a common acronym for "just in time" [to wit: compilation]. *Compilation* means generating machine-executable code; it is something that has to happen to every program for it to be runnable. The question is when.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:35]: makes sense!

File: _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd (hunk 10 above)
Comment (skeydan) [reply _posts/2021-08-10-jit-trace-module/jit-trace-module.Rmd:360]: oh, right, thanks! I think I got really confused on that one ;-)))

**Comment by skeydan** [comment]
Thanks so much @batpigandme and @dfalbel, these were super helpful suggestions!!
